scrapy/dispatch/dispatcher.py

```diff
@@ -0,0 +1,367 @@
+
+            named
+                Named arguments which will be passed to receivers.
+
+        Returns a list of tuple pairs [(receiver, response), ... ].
+        """
+        responses = []
+        if not self.receivers or self.sender_receivers_cache.get(sender) is NO_RECEIVERS:
+            return responses
+
+        for receiver in self._live_receivers(sender):
+            response = receiver(signal=self, sender=sender, **named)
+            responses.append((receiver, response))
+        return responses
+
+    def send_robust(self, sender, **named):
+        """
+        Send signal from sender to all connected receivers catching errors.
+        Modified to return Failures.
+        Arguments:
+
+            sender
+                The sender of the signal. Can be any python object (normally one
+                registered with a connect if you actually want something to
+                occur).
+
+            named
+                Named arguments which will be passed to receivers. These
+                arguments must be a subset of the argument names defined in
+                providing_args.
+        """
+        dont_log = named.pop('dont_log', _IgnoredException)
+        spider = named.get('spider', None)
+        responses = []
+        if not self.receivers or self.sender_receivers_cache.get(sender) is NO_RECEIVERS:
+            return responses
+
+        # Call each receiver with whatever arguments it can accept.
+        # Return a list of tuple pairs [(receiver, response), ... ].
+        for receiver in self._live_receivers(sender):
+            try:
+                response = receiver(signal=self, sender=sender, **named)
+                if isinstance(response, Deferred):
+                    logger.error("Cannot return deferreds from signal handler: %(receiver)s",
+                                 {'receiver': receiver}, extra={'spider': spider})
+            except dont_log:
+                response = Failure()
+            except Exception:
+                response = Failure()
+                logger.error("Error caught on signal handler: %(receiver)s",
+                             {'receiver': receiver},
+                             exc_info=True, extra={'spider': spider})
+            responses.append((receiver, response))
+        return responses
+
+    def send_robust_deferred(self, sender, **named):
+        """
+        Send signal from sender to all connected receivers catching errors.
+        Modified to work with functionons returning twisted deferreds.
+        Arguments:
+
+            sender
+                The sender of the signal. Can be anything python object (normally one
+                registered with a connect if you actually want something to
+                occur).
+
+            named
+                Named arguments which will be passed to receivers. These
+                arguments must be a subset of the argument names defined in
+                providing_args.
+        """
```

Comment:
Tiny nitpick: This docstring should probably say "Returns a Deferred that fires with a list of tuple pairs [(receiver, response), ... ]." or similar